sync/src/chain.rs

```diff
@@ -442,20 +453,67 @@ impl ChainSync {
 		self.active_peers.remove(&peer_id);
 	}
 
-	fn start_snapshot_sync(&mut self, io: &mut SyncIo, peer_id: PeerId) {
+	fn maybe_start_snapshot_sync(&mut self, io: &mut SyncIo) {
+		if self.state != SyncState::WaitingPeers {
+			return;
+		}
+		let best_block = io.chain().chain_info().best_block_number;
+
+		let (best_hash, max_peers, snapshot_peers) = {
+			//collect snapshot infos from peers
+			let snapshots = self.peers.iter()
+				.filter(|&(_, p)| p.is_allowed() && p.snapshot_number.map_or(false, |sn| best_block < sn && (sn - best_block) > SNAPSHOT_RESTORE_THRESHOLD))
+				.filter_map(|(p, peer)| peer.snapshot_hash.map(|hash| (p, hash.clone())));
+
+			let mut snapshot_peers = HashMap::new();
+			let mut max_peers: usize = 0;
+			let mut best_hash = None;
+			for (p, hash) in snapshots {
+				let peers = snapshot_peers.entry(hash).or_insert_with(Vec::new);
+				peers.push(*p);
+				if peers.len() > max_peers {
+					max_peers = peers.len();
```

Comment:
if there's a tie here it should be weighted by the best block number so you don't get very stale snapshots.